Encoders for Array and Map

jeremyrsmith opened the pull request:
* Add encoders for Array and Map
* Change Vector encoder to specialize primitives

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -27,23 +29,26 @@ abstract class TypedEncoder[T](implicit val classTag: ClassTag[T]) extends Seria
 //   - import TypedEncoder.usingDerivation
 // """)
-object TypedEncoder {
+object TypedEncoder extends TypedEncoder0 {
```
Is `TypedEncoder0` a `LowPriorityTypedEncoder`? Why do we need prioritization here?

jeremyrsmith replied:
Yes, it is for priority. I had added prioritization while trying to figure out `Map` and it seemed useful to avoid ambiguity here (when injection would essentially always be preferred over derivation if it were available).

I can revert this change if you like.

OlivierBlanvillain replied:
If it's not necessary for this PR I would rather see it revered, having ambiguities when defining things you're not supposed to is IMO a good thing :)

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,24 +337,141 @@ object TypedEncoder {
+          case FloatType => "toFloatArray" -> ScalaReflection.dataTypeFor[Array[Float]]
+          case DoubleType => "toDoubleArray" -> ScalaReflection.dataTypeFor[Array[Double]]
+        }.map {
+        case (method, typ) => Invoke(path, method, typ)
```
indentation is off

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,24 +337,141 @@ object TypedEncoder {
+    val targetDataType: DataType = DataTypes.createArrayType(underlying.targetDataType)
+
+    def constructorFor(path: Expression): Expression = {
+      Option(underlying.sourceDataType)
```
Could we factor this out with https://github.com/typelevel/frameless/pull/113/commits/d929233efd7bafad4759aee569710c87fc968064#diff-f0794dc7ae1a26e5c09857691a6b8855R293?

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,24 +337,141 @@ object TypedEncoder {
+    implicit
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[scala.collection.immutable.Map[A, B]] = new TypedEncoder[scala.collection.immutable.Map[A, B]] {
```
IIRC Spark reflection encoder is for `Map` is for `scala.collection.Map` and not `scala.collection.immutable.Map`. I wonder if we should follow (or provide both) for compatibility.

jeremyrsmith replied:
If it were for `scala.collection.Map`, you would have to ascribe that type everywhere - unless both were provided. Though maybe we could just provide `scala.collection.Map` and an injection would take care of `immutable.Map`?

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,24 +337,141 @@ object TypedEncoder {
+    }
+  }
+
+  implicit def mapEncoder[A : ClassTag, B : ClassTag](
```
Is this closing https://github.com/typelevel/frameless/issues/89?

jeremyrsmith replied:
Hard to say, link is dead in #89 - but sounds like it

jeremyrsmith commented:
@OlivierBlanvillain added a commit to address your comments - reverted the prioritization change and fixed the other issues.

I changed the `Map` encoder to be for `Map` (i.e. `scala.Predef.Map`) which is still `scala.collection.immutable.Map`. I think this is better for now, because most people will be constructing the `Map` with either the constructor from `Predef` or `toMap` on a collection. If people want to use the base `scala.collection.Map` trait, they can just call `toMap` on it (which is necessary for many reasons, including the fact that `scala.collection.Map` is not necessarily serializable).

OlivierBlanvillain commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -279,47 +278,54 @@ object TypedEncoder extends TypedEncoder0 {
+        case FloatType   => "toFloatArray"   -> ScalaReflection.dataTypeFor[Array[Float]]
+        case DoubleType  => "toDoubleArray"  -> ScalaReflection.dataTypeFor[Array[Double]]
+      }.map {
+      case (method, typ) => Invoke(path, method, typ)
```
Indentation still off

OlivierBlanvillain commented:
This is awesome! @kanterov do you want to give this another look?

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
+    val targetDataType = MapType(encodeA.targetDataType, encodeB.targetDataType, encodeB.nullable)
+
+    // needed to construct Map using `ArrayBasedMapData.toScalaMap` (avoids an extra MapObjects)
+    private val arrayA = arrayEncoder[A]
```
Do we need this as `val`? Probably the whole thing could be done as
```scala
implicit def mapEncoder[
  A : ClassTag : TypedEncoder, 
  B : ClassTag : TypedEncoder
]: TypedEncoder[Map[A, B]] = new TypedEncoder[Map[A, B]] {
...

    def constructorFor(path: Expression): Expression = {
      val keyArrayType = ArrayType(encodeA.targetDataType, false)
      val keyData = wrap(arrayEncoder[A].constructorFor(Invoke(path, "keyArray", keyArrayType)))
```

jeremyrsmith replied:
I can't explain why, but before I captured these as values, Spark's codegen would generate bad code at runtime and blow up. I'll try changing only that now and see if it was a coincidence with something else.

jeremyrsmith replied:
Actually, I think it might be better to capture them as vals.

Imagine `TypedDataset[Array[Map[Int, Int]]` - the outer encoder is going to call the inner encoder potentially thousands or millions of times during a plan execution. If we call `arrayEncoder[A]` and `arrayEncoder[B]` each time, it is going to result in a lot of needless allocation/dellocation and will potentially hurt hotspotting (since each instance of `arrayEncoder[A]` output is its own refinement class - maybe that should be changed?).

Does capturing the values do harm?

kanterov replied:
As I know, code generation in Spark happens only once, then Spark generates Java code and executes it, this code gets executed only on driver node before actual job gets executed, so there is no need to do any kind of optimizations.

jeremyrsmith replied:
Yes, you're right of course. Didn't occur to me.

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
-          val invoke = Invoke(Literal.fromObject(inj), "apply", trb.sourceDataType, Seq(path))
-          trb.extractorFor(invoke)
-        }
+  (implicit inj: Injection[A, B], trb: TypedEncoder[B]): TypedEncoder[A] =
```
Has anything changed here?

jeremyrsmith replied:
My editor keeps messing with the indentation. 🤦‍♂️

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
+    check(prop[String, Float])
+    check(prop[String, Double])
+    check(prop[String, X1[String]])
+    check(prop[String, String])
```
What about tests where:
1. key isn't `String`
2. key is `Option[_]`
3. key is `X1[_]`
4. value is `Option[_]`

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
+    check(prop[Long])
+    check(prop[Float])
+    check(prop[Double])
+    check(prop[X1[String]])
```
Would be nice to have combinations of `X1`, `Option`, `Int` and `String` as well.

kanterov replied:
There might be a corner case encoding both `Option[Array[A]]` and `Array[Option[A]]` for both primitive and non-primitive `A`.

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
     classTag: ClassTag[F]
   ): TypedEncoder[F] = new RecordEncoder[F, G]
-}
+}
```
Needed?

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
+        convertedKeys :: convertedValues :: Nil,
+        dataType = targetDataType)
+    }
+
```
Needed?

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
+    }
+
+    def constructorFor(path: Expression): Expression = {
+      val keyArrayType = ArrayType(encodeA.targetDataType, false)
```
Why nullable is always `false`?

jeremyrsmith replied:
I had assumed that we were reserving nullable for `Option` types? That seems to be the case everywhere else.

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
 
+  test("array fields") {
+
+    def prop[T](implicit arb: Arbitrary[Array[T]], encoder: TypedEncoder[X1[Array[T]]]) = forAll {
```
Probably implicits could be replaced with
```scala
def prop[T : Arbitrary : TypedEncoder]: Prop
```

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
+
+    def prop[T](implicit arb: Arbitrary[Array[T]], encoder: TypedEncoder[X1[Array[T]]]) = forAll {
+      data: Array[T] =>
+        val Seq(X1(arr)) = TypedDataset.create(Seq(X1(data))).collect().run()
```
What about a test where `Array` is root element, not field, so type would be `TypedDataset[Array[T]]`

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[Map[A, B]] = new TypedEncoder[Map[A, B]] {
+    val nullable: Boolean = false
```
What is reasoning behind using `val` instead of `def`? Asking because everywhere else we use `def`

jeremyrsmith replied:
I'll change them to `def`, I am just in the habit of using `val` for stable definitions

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
+    def prop[A, B](implicit arb: Arbitrary[Map[A, B]], encoder: TypedEncoder[X1[Map[A, B]]]) = forAll {
+      data: Map[A, B] =>
+        val Seq(X1(map)) = TypedDataset.create(Seq(X1(data))).collect().run()
+        Prop(map == data)
```
Isn't `map ?= data` better because it would generate pretty message?

kanterov replied:
By the way, there are `CollectTests` that might be a better place for this code

kanterov commented on dataset/src/test/scala/frameless/CreateTests.scala:
```diff
@@ -28,6 +28,66 @@ class CreateTests extends TypedDatasetSuite with Matchers {
+
+  test("map fields (scala.Predef.Map / scala.collection.immutable.Map)") {
+
+    def prop[A, B](implicit arb: Arbitrary[Map[A, B]], encoder: TypedEncoder[X1[Map[A, B]]]) = forAll {
```
Probably could be replaced with
```scala
def prop[A : TypedEncoder : Arbitrary, B : TypedEncoder : Arbitrary]
```

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -34,16 +35,19 @@ object TypedEncoder {
     def nullable: Boolean = true
 
     def sourceDataType: DataType = ScalaReflection.dataTypeFor[Unit]
+
```
I don't want to start a holy war, but I don't like these line breaks and they weren't here before. However, we don't have any guidelines for code style and formatting, so the only way I can argue here is minimizing the diff and keeping code consistent.

jeremyrsmith replied:
I didn't do this intentionally, no need for a holy war :)

kanterov commented on dataset/src/main/scala/frameless/TypedEncoder.scala:
```diff
@@ -296,30 +342,119 @@ object TypedEncoder {
+
+    def extractorFor(path: Expression): Expression = {
+      val keys =
+        Invoke(
```
Given generated code in on critical path, I think it completely makes sense avoid allocating `Vector` with custom functions `Iterator[Int] => Array[Int]` for primitives and `Iterator[A] => Array[A]` for boxed types.

kanterov replied:
There are `TypedEncoderUtils` where we put such things.

imarios commented:
@kanterov is so thorough : ). I wish all the people I work with gave so much time to review code and give suggestions.

imarios commented:
Hey guys, this is a great addition. The PR has remained idle for a while. Anything I can do to help here? @jeremyrsmith

imarios commented:
Hey @jeremyrsmith , any chance of getting this ready for 0.3.0? We are getting close to a new release.

iravid commented:
Anything I could do to help get this in to 0.4? I wouldn't mind addressing @kanterov's comments if that's alright with you, @jeremyrsmith.

imarios commented:
@iravid I think if you can help here it would be great!

jeremyrsmith commented:
@iravid of course, please go for it!

iravid commented:
So I have almost everything addressed, but I'm getting failures on the vector sorting test:
```
[info] - sort on vector test: descending order *** FAILED *** (385 milliseconds)
[info]   RuntimeException was thrown during property evaluation.
[info]     Message: Error while decoding: java.lang.ClassCastException: [Ljava.lang.Character; cannot be cast to [C
[info]   staticinvoke(class frameless.TypedEncoderUtils$, ObjectType(class scala.collection.immutable.Vector), mkVector, mapobjects(MapObjects_loopValue2560, MapObjects_loopIsNull2561, StringType, frameless.TypedEncoder$$anon$9$$anon$10@3116755d.invert, input[0, array<string>, false]).array, true)
[info]   +- mapobjects(MapObjects_loopValue2560, MapObjects_loopIsNull2561, StringType, frameless.TypedEncoder$$anon$9$$anon$10@3116755d.invert, input[0, array<string>, false]).array
[info]      +- mapobjects(MapObjects_loopValue2560, MapObjects_loopIsNull2561, StringType, frameless.TypedEncoder$$anon$9$$anon$10@3116755d.invert, input[0, array<string>, false])
[info]         :- frameless.TypedEncoder$$anon$9$$anon$10@3116755d.invert
[info]         :  :- frameless.TypedEncoder$$anon$9$$anon$10@3116755d
[info]         :  +- lambdavariable(MapObjects_loopValue2560, MapObjects_loopIsNull2561, StringType).toString
[info]         :     +- lambdavariable(MapObjects_loopValue2560, MapObjects_loopIsNull2561, StringType)
[info]         +- input[0, array<string>, false]
[info]   
```

It seems to be related to the character array becoming a boxed/unboxed array, or something of the sort. Would appreciate help deciphering this.

imarios commented:
@iravid is this PR still relevant? Should we close it? Also, @kanterov mentioned some performance issues with the encoders. I didn't follow what we did here. Where those concerns addressed in our current Array and Map encoders?

iravid commented:
I believe this can be closed, yes. Those perf concerns were addressed, yes- we convert maps using native Catalyst support (same as native Spark does)

imarios commented:
Hey guys, @iravid @jeremyrsmith can we close this one?
